File: src/main/java/legacy/AppliedDiscount.java
package legacy;

import java.time.Instant;
import java.util.Objects;

abstract class AppliedDiscount {

    private final Long id;
    private final Money amount;
    private final UserId userId;
    private final Instant applicationTime;

    protected AppliedDiscount(Long id, Money amount, UserId userId, Instant applicationTime) {
        this.id = id;
        this.amount = amount;
        this.userId = userId;
        this.applicationTime = applicationTime;
    }

    public Money getAmount() {
        return amount;
    }

    public UserId getUserId() {
        return userId;
    }

    public Long getId() {
        return id;
    }

    public Instant getApplicationTime() {
        return applicationTime;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AppliedDiscount that = (AppliedDiscount) o;
        return Objects.equals(id, that.id) && Objects.equals(amount, that.amount) && Objects.equals(userId, that.userId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, amount, userId);
    }
}

File: src/main/kotlin/refactored/AppliedDiscount.kt
package refactored

import java.time.Instant

sealed class AppliedDiscount() {
    abstract val id: Long
    abstract val amount: Money
    abstract val userId: UserId
    abstract val applicationTime: Instant

    class FixedDiscount(
        override val id: Long,
        override val amount: Money,
        override val userId: UserId,
        override val applicationTime: Instant
    ) : AppliedDiscount()

    class PercentageDiscount(
        override val id: Long,
        override val amount: Money,
        override val userId: UserId,
        override val applicationTime: Instant
    ) : AppliedDiscount()
}